cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,524 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cstdio>
+#include <execution>
+#include <raft/core/cuda_support.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/stream_view.hpp>
+#include <raft/core/resources.hpp>
+#include <type_traits>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/cudart_utils.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resource/cublas_handle.hpp>
+#include <raft/linalg/detail/cublas_wrappers.hpp>
+#ifdef __CUDACC__
+#include <raft/util/cuda_dev_essentials.cuh>
+#endif
+#endif
+
+namespace raft {
+namespace detail {
+
+template <bool B, typename DstType = void, typename SrcType = void, typename T = void>
+struct mdspan_copyable {};
+
+/*
+ * A helper struct used to determine whether one mdspan type can be copied to
+ * another and if so how
+ */
+template <typename DstType, typename SrcType, typename T>
+struct mdspan_copyable<true, DstType, SrcType, T> {
+  using dst_type = std::remove_reference_t<DstType>;
```

Comment:
Can we add `static_assert` here to ensure the types are indeed mdspans, and if not we pass a clear error message to the user? Otherwise there will be a bunch of error messages that will be hard to parse

Reply:
In general, I'm not too concerned about this. `raft::detail::mdspan_copyable` is an implementation detail of `raft::copy` and should never be consumed outside of the context of `raft::copy`. That will be SFINAE'd out for any instantiation that does not take mdspan arguments. If we want to add a `static_assert` for debugging `mdspan_copyable` itself, I'm not opposed, but I don't think it adds anything to the actual safety of the implementation. The compilation error for consumers will still be a SFINAE failure.

Reply:
Consider the scenario where I called `raft::copy(p)` where `p` is of type `int*`. Even though `mdspan_copyable` is an implementation detail, the user will get a bunch of errors from all the code internally that tries to access mdspan internals such as `mdspan::layout_type`, etc. `static_assert` will help us keep the error tree clean.

Reply:
Here's a sketch of an overload and error-checking pattern.

```c++
// Outer interface
template<class InView, class OutView>
void copy(InView in, OutView out) {
  // Check types of in and out individually, and transform each into a "canonical" type.
  // For example, if InView is mdspan<T, ...>, transform to `mdspan<add_const_t<T>, ...>`.
  auto in_canon = detail::canonicalize(in);
  auto out_canon = detail::canonicalize(out);

  // Check that decltype(in_canon) can be copied to decltype(out_canon).
  // Each type has already been checked and canonicalized.
  // For example, we know that out_canon is an mdspan, span,
  // or raw pointer to nonconst.  detail::copyable would check,
  // e.g., that two mdspans have the same ranks and that any
  // extents known at compile time match.
  //
  // We never call detail::copyable for noncanonical types,
  // so it doesn't need to check the input types individually,
  // and it only needs overloads for canonical types.
  static_assert(detail::copyable(in_canon, out_canon));

  // Any run-time precondition checks, e.g., for run-time extents.
  assert(detail::runtime_copyable(in_canon, out_canon));

  // At this point, we know that in_canon can be copied to out_canon.
  // We will never call detail::copy for any types other than
  // individually canonical and mutually copyable types.
  // Thus, we don't need to check errors at compile time.
  // We've also checked run-time preconditions above,
  // so we don't need to check them inside detail::copy.
  // detail::copy can just Do The Copying.
  detail::copy(in_canon, out_canon);
}
```

Reply:
I really like @mhoemmen's approach, it would simplify `detail::mdspan_copyable` by splitting it into multiple type-checkers.

Reply:
Unfortunately, that solves a different (albeit related) problem than `mdspan_copyable` is intended to solve. To give a bit more insight, let me run you through exactly why we need `mdspan_copyable` and how we got to the current design.

First off, we need a way to SFINAE-separate the instantiations that require nvcc from those that down. That's what allows us to correctly construct the public cuh and hpp files, which is important both for enabling usage in non-CUDA builds and for speeding up compilation in CUDA-enabled builds.

Already, that requires all of the following checks:

- Are we in a CUDA-enabled build?
- Are both mdspans device-accessible? If not, is one _only_ device-accessible and the other _only_ host-accessible?
- Are both views on contiguous memory?
- Do both have the same layout? If not, are they actually equivalent layouts spelled differently (e.g. 1D layout_left vs 1D layout_right)?
- Do they both have the same `value_type`?
- If they are not the same layout, are we copying a 2D layout_left to a 2D layout_right?
- If so, are both views on float or both views on doubles?

We need all that information just to correctly set up that top-level SFINAE, because that is what is required to answer "Does this copy involve the device but cannot be performed with cudaMemcpyAsync or cublas?"

Once we've SFINAE-separated the instantiations that require nvcc from those that don't, we still need to figure out exactly which copy implementation we're actually going to use. For deciding if we're going to use cudaMemcpyAsync or cublas, we need to answer all the above questions again, so we have a few possible implementations:

**Possibility 1:** SFINAE-separate each of the `copy` implementations based on the method used.

This path substantially increases the complexity of the code, but it also _massively_ increases compile time. Moving from a single outermost SFINAE to a whole tree of SFINAEs requires the compiler to evaluate a whole bunch more possible instantiations, each with their own (expensive) SFINAE condition. When I tried this, it took so long to compile even with trivial function bodies that I immediately started looking for another path.

**Possibility 2:** Perform all necessary checks for the outermost SFINAE, and then re-evaluate the necessary information in each constexpr if condition.
This makes the code significantly less readable and also increases compile time somewhat. I had assumed the compiler would cleverly handle the redundant checks, but given the actual compile times I was seeing, that does not seem to be the case.

**Possibility 3:** Separate `mdspan_copyable` into several separate objects and `std::conjunction/std::disjunction` them together.
This had a lot of appeal for me, and I tried several different ways to get it to work as desired. It's the equivalent of breaking a complex function down into many smaller well-named functions. Unfortunately, for the version that gets semi-decent compile time, we actually hit the same problem that @divyegala noticed right at the top of this thread over and over again in separate objects, each of which have to be SFINAE-protected. This significantly increases complexity, requires some redundant checks, and ultimately gives mediocre compile time.

**Final version:** What I ultimately landed on was a single object with clear, well-named type aliases and static constexpr values which could both be used for the necessary SFINAEs and re-used in the constexpr if conditions. From the compiler's perspective, the nice thing about this is that it's easy to re-use that template instantiation everywhere it's needed in the translation unit, and we perform all necessary checks for a specific instantiation exactly once. From the developer's perspective, though there's a lot going on in this object, we can follow the logic through the naming of the constexpr values. E.g. We see that the original `raft::copy` (equivalent to cudaMemcpyAsync) is used when:

1. CUDA is enabled for the build overall
2. The mdspans contain the same dtype
3. Both mdspans are contiguous
4. Both mdspans have the same underlying layout

Furthermore, we just need one top-level SFINAE check to protect everything correctly and avoid weird, intractable compiler errors. As @divyegala pointed out, I did miss something in that check, but that has been corrected in the version I just pushed. Now any `raft::copy` instantiation that we can't support gives a nice, familiar, and succinct SFINAE failure from the compiler.

If there's some other way to structure this that is more readable without messing with compile time, I'd be open to a PR to my branch, but I've already tried a bunch of different approaches, so I have no confidence I could come up with something better myself. I'd be even happier to see this as a follow-up PR to clean up implementation details if we're generally happy with the runtime, compile time, and public API. We'll need a follow-up to reach SoL copies for all cases anyway, so this can continue to evolve as we use it.

Reply:
Thanks for the answer @wphicks that details everything you tried and why you arrived at this design. Your reasons for your current approach seem more than satisfactory to me now.

Could you show the error stacktrace that the compiler throws when we try to call this function with a non-mdspan object or with a const destination?

Pending what that would look like and response to a couple of other open questions, I think we are very close to finalizing this PR.

Reply:
Thanks very much! Really appreciate your help in thinking through this so carefully. Here's a gist where we attempt a copy between two ints: https://gist.github.com/wphicks/668681c85bf59c1d565bcaa0f20ed74b, and here is a gist where we attempt to copy into a mdspan of consts: https://gist.github.com/wphicks/2c791336955a33f9f74dbb50629e2fd4

Also useful to note is that my static analyzer correctly recognizes that these are not viable overloads (unsurprising but still very satisfying!) and offers a nice warning before I even attempt to compile.

Does all that seem okay to you? @divyegala, feel free to hit the "Resolve conversation" button if you're satisfied; I don't want to cut off discussion prematurely.

Reply:
The error gists look great! Thank you for these, this is looking really good.